operators/grpc_operators/python/common/tensor_proto.py

```diff
@@ -0,0 +1,215 @@
+
+
+class ProtoTensorCudaArrayInterface:
+    def __init__(self, proto_tensor):
+        data = np.frombuffer(proto_tensor.data, dtype=TensorProto._get_numpy_type(proto_tensor))
+        data = cp.asarray(data)
+        self.__cuda_array_interface__ = {
+            "version": 2,
+            "shape": tuple(proto_tensor.dimensions),
+            "data": (data.data.ptr, False),
+            "typestr": proto_tensor.type_str,
+            "strides": tuple(proto_tensor.strides),
+        }
+
+
+class DLDataTypeCode(Enum):
+    kDLInt = 0
+    kDLUInt = 1
+    kDLFloat = 2
+    kDLOpaqueHandle = 3
+    kDLComplex = 5
+
+
+class TensorProto:
+    NULL_KEY_NAME = "__NULL__"
+
+    STORAGE_TYPES = {
+        holoscan_proto.Tensor.MemoryStorageType.kHost: DLDeviceType.DLCPU,
+        holoscan_proto.Tensor.MemoryStorageType.kDevice: DLDeviceType.DLCUDA,
+        holoscan_proto.Tensor.MemoryStorageType.kSystem: DLDeviceType.DLCUDAHOST,
+    }
+
+    PRIMITIVE_TYPES = {
+        holoscan_proto.Tensor.PrimitiveType.kUnsigned8: np.uint8,
+        holoscan_proto.Tensor.PrimitiveType.kUnsigned16: np.uint16,
+        holoscan_proto.Tensor.PrimitiveType.kFloat32: np.float32,
+    }
+
+    @staticmethod
+    def add_tensor_to_proto(gxf_entity, response: holoscan_proto.EntityResponse):
+        TensorProto._gxf_tensors_to_proto_tensors(gxf_entity, response.tensors)
+
+    @staticmethod
+    def tensor_to_proto(gxf_entity: Dict) -> holoscan_proto.EntityRequest:
+        entity_request = holoscan_proto.EntityRequest()
+        TensorProto._gxf_tensors_to_proto_tensors(gxf_entity, entity_request.tensors)
+        return entity_request
+
+    @staticmethod
+    def proto_to_tensor(
+        entity_request: Union[holoscan_proto.EntityRequest, holoscan_proto.EntityResponse], context
+    ) -> Entity:
+        gxf_entity = Entity(context)
+        TensorProto._proto_tensors_to_gxf_tensors(entity_request, gxf_entity)
+        return gxf_entity
+
+    @staticmethod
+    def _gxf_tensors_to_proto_tensors(gxf_entity, tensors):
+        for key, gxf_tensor in gxf_entity.items():
+            try:
+                tensor = tensors[key if key else TensorProto.NULL_KEY_NAME]
+
+                for dim in gxf_tensor.shape:
+                    tensor.dimensions.append(dim)
+
+                for stride in gxf_tensor.strides:
+                    tensor.strides.append(stride)
+
+                tensor.primitive_type = TensorProto._get_gxf_type(gxf_tensor)
+                tensor.type_str = TensorProto._get_numpy_type_str(gxf_tensor)
+                tensor.memory_storage_type = TensorProto._get_proto_storage_type(gxf_tensor)
+
+                if tensor.memory_storage_type == holoscan_proto.Tensor.MemoryStorageType.kHost:
+                    tensor.data = gxf_tensor.cpu().numpy().tobytes()
+                elif (
+                    holoscan_proto.Tensor.MemoryStorageType.kDevice
+                    or holoscan_proto.Tensor.MemoryStorageType.kSystem
```

Comment:
why is the host-side "kSystem" case down here in the device array code path and not the host one above?